### PR TITLE
Refactored GitHub Actions

### PR DESCRIPTION
**Pull request by smithdc1**
The GitHub Actions were written some time ago. There's clearly a better way to do this so I've reworked them in the style of @adamchainz. Adam -- I hope you don't mind the plagiarism!

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        ENV_PREFIX=$(tr -C -d "0-9" <<< "${{ matrix.python-version }}")
+        TOXENV=$(tox --listenvs | grep "^py$ENV_PREFIX" | tr '\n' ',') tox
+    - name: Run lint
+      if: ${{ matrix.python-version == '3.9' }}
+      run: |
+        tox -e lint
+    - name: Upload coverage to Codecov
+      uses: codecov/codecov-action@v1
+      with:
+        token: ${{ secrets.CODECOV_TOKEN }}
```
Comment (adamchainz) [inline .github/workflows/main.yml:52]: public repos don't need the codecov token

File: tox.ini [hunk 2]
```diff
@@ -2,17 +2,27 @@
...
+    -rrequirements/testing.txt
 allowlist_externals = make
-commands = make test
+commands = python -W error::DeprecationWarning -W error::PendingDeprecationWarning -m pytest {posargs} --cov=crispy_forms
 ignore_outcome =
     django-latest: True
 setenv =
     PYTHONDONTWRITEBYTECODE=1
+
+[testenv:lint]
```
Comment (adamchainz) [inline tox.ini:21]: don't add a `lint` environment - just run `pre-commit` in a github action?

See e.g. https://github.com/django/daphne/pull/365

Comment (carltongibson) [reply tox.ini:21]: Or maybe have the lint environ use pre-commit... 🤔

(My muscle memory is objecting to not having it already. :D)

Comment (adamchainz) [reply tox.ini:21]: Well up to you

Comment (smithdc1) [reply tox.ini:21]: Now I'm really unsure! I also like the idea of pre-commit.ci 🤷‍♂️ 

I guess let's have this as it is, this is significantly better than what we have at the moment.

**Pull request merged**